### PR TITLE
my initial commit of a suggestion seed project

### PR DESCRIPTION
**Pull request by JesterXL**
Ladies & gents, here is my suggestion for a seed project.
# Installation
1. check out code
2. switch to `jesse-suggestion` branch
3. in your Terminal, cd to the directory and run `npm install && bower install`
4. once complete type `gulp`
5. open your browser to `http://localhost:8628`
# Description

Code is Angular 1.3, Gulp build system (supports only Development for now, no prod, nor does it link to our proposed Gulp build project).

**src/client**: all Angular front-end code.

**src/api**: all [Restify](http://mcavage.me/node-restify/) API related code. Open to another middleware like Hapi, no strong feelings here.

**src/static**: all static files via [Express](http://expressjs.com/). Seperate from API mainly because you optimize static servers differently, and because static file logic is very different from API (deeplinking, caching, cdn, etc).

TODO:
- add Redis caching layer example
- add Mongo data layer
- add Passport OAuth example
# Gulp Tasks

`gulp`

Deploys your app locally, starts the static & API server. Connect to `http://localhost:8628` to see it.

`gulp analyze`

Shows formatting, styling, and complexity (Halsted, Cyclomatic) in your code.

`gulp analyzeWhileIFix`

Same as analyze, but will fail unless all 3 quality gates are passed. That way, you can see what's broke, fix it, and get all 3 passing vs 1 big list of text in the Terminal.

`gulp test`

Runs Karma unit tests and generates code coverage reports for both client and CI server.

`gulp testWhileICode`

Same as Karma, but runs in CI mode. When you save a file, it'll rerun all the unit tests wicked fast.

`gulp showCoverage`

Shows the code coverage report generated from `gulp test`.

`gulp judge`

The ultimate quality gate. If you pass all of these, you code is epic tier and ready for Git push / SVN commit.

`gulp cow`

Test that you have native Promise support in your version of Node. If not, use Bluebird.

### CODE REVIEW

File: src/client/main/macroTarget.factory.js [hunk 1]
```diff
@@ -0,0 +1,159 @@
+/* global _ */
+(function() {
```
Comment (tomi-vanek) [inline src/client/main/macroTarget.factory.js]: Naming conventions: file names are typically dasherized: macro-target.factory.js

File: src/api/testapi.js [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+});
+
+api.get('/testmongo', function(req, res, next)
+{
+    console.log("testing MongoClient...");
+    testMongo();
+    res.send(200, {response: true});
+});
+
+function testMongo()
```
Comment (tomi-vanek) [inline src/api/testapi.js]: I would not put any backend / DB into the API... makes the solution unnecessary complex.

**Comment by thomaswmanion** [comment]
Cool!

Any reason to wrap everything in an anonymous function and then not define the Angular component function anonymously? I see a lot of people doing it but I'm not too crazy about it myself... Maybe I just don't get it.

```
(function() {
  'use strict';
  module.controller('MyController', MyController);
  function MyController(){}
})();
```

vs.

```
'use strict';
module.controller('MyController' function() {});
```

**Comment by achristoph** [comment]
I think it looks cleaner when there are multiple components declaration in one file - declarations on the top, functions on the bottom, but if there's only one declaration per file, I prefer the anonymous function - it saves some spaces and characters that might add up to significant size when the app gets bigger.

**Comment by tomi-vanek** [comment]
Protection of global context in case somebody forgets “var” for declaration of variables.
In case you have distributed team of unexperienced developers, the style and idioms tend to be more defensive ;-)
tomi

File: src/client/main/macroTarget.factory.js (hunk 1 above)
Comment (JesterXL) [inline src/client/main/macroTarget.factory.js]: As long as we agree to go full-OOP in Angular 2 and follow the camel case standard, I'll survive.

File: src/api/testapi.js (hunk 2 above)
Comment (JesterXL) [inline src/api/testapi.js]: K, took out Mongo, left in Redis for caching.

**Comment by JesterXL** [comment]
Made updates based on comments. Need at least 4 approvers. Either "Approve", or "fix/change this".

**Comment by JesterXL** [comment]
Also, once approved I'll rip out Gulp, put in the Gulp repo, then merge into this one as an npm dependency.

**Comment by JesterXL** [comment]
@achristoph The reason we separate out, and I reckon John Papa recommends it too ( https://github.com/johnpapa/angular-styleguide ) is that multiple definitions in a file don't stay that way. As your code grows in size, the whole file grows. If they are separate, you can keep the lines of code down and more readable vs. 1,000 lines of code in a single file. It also makes it easier to merge that 1 file in a group environment since you're just comparing 100 lines of code or less against another file of same size that is just a Controller for example.

Regarding anonymous, I agree. To keep code coverage up, you only use anonymous functions on an object so you can target and test it. If you use function declaration, while they are "private", you can't easily target them in tests, and code coverage cites that as a missing test block which isn't good.

**Comment by JesterXL** [comment]
BTW, if you're not an owner/invited, lemme know. I invited you @achristoph

**Comment by JesterXL** [comment]
Also, last comment on multiple files; it's basically analogous to how Angular 2 works with each file being a class.

So these 3 files:

``` javascript
// File main.module.js
angular
    .module('main.about', [
        'ui.router',
        'LocalStorageModule'
        ]);

// File main.directive.js
angular
    .module('main.about')
    .directive('fsAbout', fsAbout);

function fsAbout()
{
    return {
        restrict: 'E',
        scope: {},
        transclude: false,
        templateUrl: 'main/about/about.directive.html',
        controller: 'fsAboutController',
        controllerAs: 'vm'
    };
}

// File main.controller.js
angular.module("main.about")
    .controller("fsAboutController", fsAboutController);

/* @ngInject */
function fsAboutController($scope, $http, localStorageService)
{
    var vm = this;
    vm.title = "Loading...";
    vm.content = "";
    $http.get('http://'+window.location.hostname+':2146/api/abouttext')
            .success(function(response)
            {
                localStorageService.set('abouttext', response.data);
                vm.title = "About";
                vm.content = response.data;
            })
            .error(function(error)
            {
                console.error(error);
                vm.title = "About";
                vm.content = "Couldn't get data from server to show you what this is about.";
            });
}
```

Becomes something like this:

``` javascript
// MainApp.js
import {Component, View, bootstrap} from 'angular2/angular2';
import {AboutComponent} from 'AboutComponent.js';

// AboutComponent.js
import {Http} from 'angular2/whateverhttpis';
import {localStorageService} from 'angular2/plugins/localStorageService';

@Component({
  selector: 'fs-about'
})
@View({
  templateUrl: 'AboutComponent.html'
})
class AboutComponent
{
  title:string;
  content:string;

  constructor()
  {
    this.name = '';
    this.content = '';
    var vm = this;
    Http.get('http://'+window.location.hostname+':2146/api/abouttext')
    .success(function(response)
    {
        localStorageService.set('abouttext', response.data);
        vm.title = "About";
        vm.content = response.data;
    })
    .error(function(error)
    {
        console.error(error);
        vm.title = "About";
        vm.content = "Couldn't get data from server to show you what this is about.";
    });
  }
}
```

**Comment by thomaswmanion** [comment]
I approve.

**Comment by JesterXL** [comment]
Alright @egshankle  @achristoph @tomi-vanek and  @jpshankle , just need 3 out of 4 of you to approve.

**Comment by egshankle** [comment]
Approve :)

**Comment by achristoph** [comment]
@JesterXL . yup. got the invite. 
cool. makes sense. thanks for the angular 2.0+ES6 intro :)
should we make use of gulp's require-dir  or is that something can be added later when the app is getting more complex if it is recommended at all?
I approve :+1:

**Comment by JesterXL** [comment]
That's 3, 1 more...

**Comment by JesterXL** [comment]
@achristoph Yeah, good point, we should. Is that the same as grunt-task where it removes all the stupid grunt.loadNpmTask calls, except require in Gulp? Yeah, that's a PR we should add. I'll add to the list of tasks tomorrow morning.

File: gulpfile.js [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+// **********************************************************************
+
+// tells you everything wrong with your code, quickly. Format, styling, and complexity.
+gulp.task('analyze', function() {
+  return gulp.src(CONFIG.client.sourceFiles)
+    .pipe(jshint())
+    .pipe(jshint.reporter('jshint-stylish'))
+    .pipe(jshint.reporter('fail'))
+    .on('error', function(e)
+    {
```
Comment (tomi-vanek) [inline gulpfile.js:45]: this is the last one :)
We have to agree explicitly of style guide, that determines also such subtle things like starting { in the same line with command, or on the next line. There are more establishing styles, or we can agree on our own. I would prefer to pick one existing.

**Comment by tomi-vanek** [comment]
One comment on style guides, otherwise APPROVED.

**Comment by achristoph** [comment]
@JesterXL . yep, that's the one

**Comment by JesterXL** [comment]
@tomi-vanek Roger that, we can automate most of these styles so if lazy bastards like me don't want to follow it, gulp can fix, regardless of IDE.

T3H MERGING

**Pull request merged**